### ISSUE TITLE
Extensibility API for niche calling

### ISSUE THREAD

**Issue by grst** [scverse/squidpy#1285]
### Description of feature

Hi @selmanozleyen, @shashkat and @Intron7,

in https://github.com/scverse/squidpy/pull/1245 a new, modular framework for niche calling was implemented. While I already provided implementation-specific feedback in https://github.com/scverse/squidpy/issues/1277, here I want to propose an "extensibility API" that allows to build custom niche calling algorithms based on the existing building blocks, similar to what's already implemented [for the neighborhood algorithms](https://squidpy.readthedocs.io/en/stable/extensibility.html). 

That means defining an interface an "embedder" and a "clusterer" and a `calculate_niche_custom` function as a high level API (analogous to [`spatial_neighbors_from_builder`](https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.spatial_neighbors_from_builder.html#squidpy.gr.spatial_neighbors_from_builder). Compared to the current implementation, I would increase the level of abstraction. An embedder or clusterer doesn't need to know anything about the AnnData object or its structure, it just gets array(s) as input and produces array(s) as output. That makes them easier to test and reason about and increases the compatibility with the sklearn ecosystem (see below). 

## Clusterer

A clusterer takes an embedding and generates labels. Instead of rolling our own interface, I think we should piggyback on scikit-learn. This is nice, because (1) it's a well-known API and (2) we can just use off-the-shelf scikit-learn algorithms without any custom code! 

Anything that implements the `ClusterMixin` interfaces qualifies:
 * `_GMMClusterer` goes away entirely, just use `sklearn.mixture.GaussianMixture`. 
 * Want to use `kmeans` for clustering for its speed? Use [`sklearn.cluster.KMeans`](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html)! 
 * Want to try BayesianGMM with cellcharter? Use [`sklearn.mixture.BayesianGaussianMixture`](https://scikit-learn.org/stable/modules/generated/sklearn.mixture.BayesianGaussianMixture.html#bayesiangaussianmixture)! 
 * or implement your own (in our case, leiden)
   ```python
   class LeidenClusterer(ClusterMixin, BaseEstimator):
       def __init__(self): ... # set parameters
       def fit(self, X, y=None): 
            tmp_ad = ad.AnnData(X=X)
            sc.pp.neighbors(tmp_ad)
            sc.tl.leiden(tmp_ad)
            return tmp_od.obs["leiden"]
    ```
 
## Embedder

This is essentially a `Transformer` in the scikit-learn world, but it's a bit trickier, because in addition to a feature space, our Embedders typically also require a spatial neighborhood graph. Relying on the scikit-learn API directly doesn't cover us. So I'd suggest to go with something that is still sklearn-like: 

```python
class BaseSpatialTransformer(BaseEstimator):
    def fit_transform(self, X, y=None, *, S=None, **fit_params):
        if y is None:
            return self.fit(X, S=S, **fit_params).transform(X, S=S)
        return self.fit(X, y, S=S, **fit_params).transform(X, S=S)
    
    def fit(self, X, y=None, *, S=None): ...
    def transform(self, X, *, S=None): ...
```

Additionally, one could look into the experimental [metadata routing](https://scikit-learn.org/stable/metadata_routing.html) which pass the spatial matrix along when using scikit-learn features such as `Pipeline()` or `GridSearchCV()`, but I don't think it's that relevant at this point. 

## Piecing it together

Moving out (1) extraction of the matrics from AnnData and (2) hyperparameter search from the Embedders and Clusterers means that the outer functions need to do a bit more. I still think it's better, as it makes the individual components more reusable. I think the solution could be that instead of a clusterer and an embedder class, "recipe" callback function is passed to the `calculate_niche_custom` function that takes an AnnData slice and returns one or multiple `.obs` columns. Alternatively, there could be another base class that defines an abstract function for this. 

For `calculate_niche_neighborhood` this would look somewhat like this

```python
def calculate_niche_neighborhood(adata, groups, *, mask, library_key, resolutions, spatial_connectivities_key, n_neighbors, scale, distance, abs_nhood):
    embedder = NeighborhoodProfileEmbedder(scale=scale, distance=distance, abs_nhood=abs_nhood)
    clusterer = LeidenClusterer(n_neighbors=n_neighbors, ...)
    
    def _recipe(adata_slice):
         embedding = embedder.fit_transform(X = adata.obs[groups], S = adata.obsp[spatial_connectivities_key])
         obs = {}
         for res in resolutions:
             obs[f"leiden_{res}"] = clusterer.fit_predict(embedding)
        return obs
   
    run_niche_pipeline(adata, maks=mask, library_key=library_key, recipe=_recipe)
```

Implementing a "custom" niche calling algorithm, e.g. combining neighborhood embedding with GaussianMixture could be easily realized that way. 

Let me know what you think!

**Comment by selmanozleyen** [comment]
Thanks I will do some prototyping on this. But if we were to expose clusterers etc. I think we shouldn't expose them under `sq.tl.*`. I think they should belong in `sq.tl.clusterers` or something like that

**Comment by grst** [comment]
> I think we shouldn't expose them under sq.tl.*. I think they should belong in sq.tl.clusterers or something like that

fully agree, same system as with neighborhood algorithms.

**Comment by selmanozleyen** [comment]
Ok, can you have also a look at the https://github.com/scverse/squidpy/pull/1287 stack perhaps? It's kind of related because it wouldn't expose the spatial_neighbours classes under sq.tl.*, which looks ugly and unstructured.

<img width="310" height="593" alt="Image" src="https://github.com/user-attachments/assets/9247bcc6-5a68-4ffa-9559-68321cd3bdc2" />

After the PR only under sq.tl.neighbours they will be exposed

**Comment by selmanozleyen** [comment]
Btw, I think I would hold the class based exposure of embedders because there is only one consumer for each case from what I understand right? At least the clusterers would be reused in cluster autok

**Comment by grst** [comment]
>  there is only one consumer for each case from what I understand right

In the internal implementation, yes. But the whole point of this would be that one can quickly build alternative niche calling algorithms using the existing building blocks.

**Comment by selmanozleyen** [comment]
Ok what about this?

```python
def nhood_aggregate(
    adata, *, groups=None, use_rep=None, layer=None,
    connectivity_key="spatial_connectivities",
    hops=(1,),                    # (1,) utag | (1..k) neighborhood | (0..k) cellcharter
    hop_mode="power",             # "power" = A@A | "shell" = excludes revisits
    combine="concat",             # "concat" | "sum"
    hop_weights=None,
    aggregation="mean",           # "mean" | "variance"
    key_added="X_nhood", copy=False,
) -> AnnData | None
```

Then the embedders become only this:

| flavor | derivation |
|---|---|
| `neighborhood` | `nhood_aggregate(groups=..., hops=(1..k), hop_mode="power", combine="sum", hop_weights=w)` → `sc.pp.scale` |
| `utag` | `nhood_aggregate(hops=(1,))` → `sc.pp.pca` |
| `cellcharter` | `nhood_aggregate(hops=(0..k), hop_mode="shell", combine="concat", aggregation=...)` → `sc.pp.pca` |

I'd like to really justify if we are going to expose. Because "embedder" sounds too generic and I am worried it would be only reused in the scope of having a custom calculate niche. Clusterer makes sense because it fits well with sklearn.   
We also need to think about the potential rsc compat issues:
https://github.com/scverse/squidpy/pull/1245#pullrequestreview-4898736406

also related: https://github.com/scverse/squidpy/issues/784

**Comment by shashkat** [comment]
Thanks for the suggestions @grst! My inputs about this are the following:
- I really like the idea of making the clusterers sklearn compatible. I cant see much downside to that, and many possible benefits.
- However, I believe for embedders it may be more tricky. Afaik, different embedders can require more/other information than just the spatial graph and annotations. For example, UTAG also needs the expression matrix, and in the future there may be embedders which may need other information from the adata object.
- This limitation of embedders needing more information (and possibly other information in the future) helps us in one way atleast - we encode the logic of obtaining the embedding given the adata object completely in the embedder object, and that allows us to design `calculate_niche_custom` in a way so as to obviate the need for the end user to specify a "recipe".

**Comment by grst** [comment]
I think there could indeed be some benefit from splitting up user-facing functions and implementing something like `nhood_aggregate`. It would be akin with generating embeddings in the single-cell world and they turn out to be useful for other things than just clustering (e.g. direct visualiztion of the embedding, batch correction, label transfer). 

In the single-cell world we have the workflow

```python
sc.tl.pca() # create embedding, can also be scVI or whatever
sc.tl.neighbors() 
sc.tl.leiden()
```

In the spatial world we currently have

```python
sq.tl.calculate_niche()
```

But maybe we should have (at least as an option)

```python
sq.tl.neighborhood_aggregate() # generate embedding
sq.tl.neighbors()
sq.tl.leiden() 
# or use a clustering algorithm without neighbors, e.g. GMM, 
sq.tl.cluster(X = A.obsm["nhood_aggregate"], method : Literal["gmm", "kmeans"] | sklearn.ClusterMixin ="gmm")
```


About your concrete function signature for `nhood_aggregate`, I'm wondering if we are not again lumping too much functionality into one function (which was the main reason for #1245 in the first place). But it's probably still ok, because unlike the old `calculate_niche`, it doesn't have mutually exclusive arguments. 

I don't think #784 is that related, because here we still create a new aggregation for each observation, while in #784 the goal is to reduce the observations to biological replicates. Or where exactly would you have seen synergies?

**Comment by grst** [comment]
> However, I believe for embedders it may be more tricky. Afaik, different embedders can require more/other information than just the spatial graph and annotations. For example, UTAG also needs the expression matrix, and in the future there may be embedders which may need other information from the adata object.

I also thought about this and it might be the better solution. Which goes into the same direction as `neighborhood_aggregate`.

**Comment by selmanozleyen** [comment]
> About your concrete function signature for nhood_aggregate, I'm wondering if we are not again lumping too much functionality into one function

I agree but if we do nhood_aggregate vs nhood_concatenate it would be more natural split. We don't need to expose it but it's something to consider I'd say
